File: lingting-spring-web/src/main/java/live/lingting/spring/web/converter/StringToLocalDateTimeConverter.java
package live.lingting.spring.web.converter;

import live.lingting.framework.util.LocalDateTimeUtils;
import org.springframework.core.convert.TypeDescriptor;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.Collections;
import java.util.Set;

/**
 * @author lingting 2022/9/28 12:17
 */
@Component
public class StringToLocalDateTimeConverter implements Converter<LocalDateTime> {

	@Override
	public boolean matches(TypeDescriptor sourceType, TypeDescriptor targetType) {
		return String.class.isAssignableFrom(sourceType.getType())
				&& LocalDateTime.class.isAssignableFrom(targetType.getType());
	}

	@Override
	public Set<ConvertiblePair> getConvertibleTypes() {
		return Collections.singleton(new ConvertiblePair(String.class, LocalDateTime.class));
	}

	@Override
	public LocalDateTime convert(Object source, TypeDescriptor sourceType, TypeDescriptor targetType) {
		String string = (String) source;
		return LocalDateTimeUtils.parse(string);
	}

}

File: lingting-spring-web/src/main/kotlin/live/lingting/spring/web/converter/StringToLocalDateTimeConverter.kt
package live.lingting.spring.web.converter

import java.time.LocalDateTime
import live.lingting.framework.util.LocalDateTimeUtils
import org.springframework.core.convert.TypeDescriptor
import org.springframework.core.convert.converter.GenericConverter.ConvertiblePair
import org.springframework.stereotype.Component

/**
 * @author lingting 2022/9/28 12:17
 */
@Component
class StringToLocalDateTimeConverter : Converter<LocalDateTime> {
    override fun matches(sourceType: TypeDescriptor, targetType: TypeDescriptor): Boolean {
        return String::class.java.isAssignableFrom(sourceType.type)
                && LocalDateTime::class.java.isAssignableFrom(targetType.type)
    }

    override fun getConvertibleTypes(): MutableSet<ConvertiblePair> {
        return mutableSetOf<ConvertiblePair>(ConvertiblePair(String::class.java, LocalDateTime::class.java))
    }

    override fun convert(source: Any?, sourceType: TypeDescriptor, targetType: TypeDescriptor): LocalDateTime {
        val string = source as String
        return LocalDateTimeUtils.parse(string)
    }
}
